Mathlib/Analysis/Convex/Side.lean

```diff
@@ -829,3 +829,219 @@ theorem isPreconnected_setOf_sOppSide (s : AffineSubspace ℝ P) (x : P) :
+      (Finset.univ.affineCombination R s.points w₁)
+      (Finset.univ.affineCombination R s.points w₂) := by
+  let w₃ : Fin (n + 1) → R := lineMap w₁ w₂ (w₁ i / (w₁ i - w₂ i))
+  have hp : 0 < w₁ i - w₂ i := by grind
+  have hw₃ : ∑ j, w₃ j = 1 := by
+    simp [w₃, lineMap_apply, Finset.sum_add_distrib, ← Finset.mul_sum, hw₁, hw₂]
+  have h : Sbtw R w₁ w₃ w₂ := sbtw_lineMap_iff.2
+    ⟨(by grind), div_pos hs₁ hp, (div_lt_one hp).2 (by grind)⟩
+  have h' : Sbtw R (Finset.univ.affineCombination R s.points w₁)
+      (Finset.univ.affineCombination R s.points w₃)
+      (Finset.univ.affineCombination R s.points w₂) := by
+    refine ⟨h.wbtw.map _, fun he ↦ ?_, fun he ↦ ?_⟩
+    · replace he := (affineIndependent_iff_eq_of_fintype_affineCombination_eq R s.points).1
+        s.independent w₃ w₁ hw₃ hw₁ he
+      simp_rw [← he, w₃] at hs₁
+      simp only [lineMap_apply, vsub_eq_sub, Pi.vadd_apply', Pi.smul_apply, Pi.sub_apply,
+        smul_eq_mul, vadd_eq_add] at hs₁
+      rw [← neg_sub (w₁ i) (w₂ i), mul_neg, div_mul_cancel₀ _ hp.ne'] at hs₁
+      simp at hs₁
+    · replace he := (affineIndependent_iff_eq_of_fintype_affineCombination_eq R s.points).1
+        s.independent w₃ w₂ hw₃ hw₂ he
+      simp_rw [← he, w₃] at hs₂
+      simp only [lineMap_apply, vsub_eq_sub, Pi.vadd_apply', Pi.smul_apply, Pi.sub_apply,
+        smul_eq_mul, vadd_eq_add] at hs₂
+      rw [← neg_sub (w₁ i) (w₂ i), mul_neg, div_mul_cancel₀ _ hp.ne'] at hs₂
+      simp at hs₂
+  refine h'.sOppSide_of_notMem_of_mem ?_ ?_
+  · intro h
+    rw [range_faceOpposite_points] at h
+    exact hs₁.ne' (s.independent.eq_zero_of_affineCombination_mem_affineSpan hw₁ h
+      (Finset.mem_univ i) (by simp))
+  · rw [range_faceOpposite_points]
+    refine affineCombination_mem_affineSpan_image hw₃ ?_ s.points
+    simp only [Finset.mem_univ, Set.mem_compl_iff, Set.mem_singleton_iff, Decidable.not_not,
+      lineMap_apply, vsub_eq_sub, vadd_eq_add, Pi.add_apply, Pi.smul_apply, Pi.sub_apply,
+      smul_eq_mul, forall_const, forall_eq, w₃]
+    rw [← neg_sub (w₁ i) (w₂ i), mul_neg, div_mul_cancel₀ _ hp.ne']
+    simp
+
+lemma sSameSide_affineSpan_faceOpposite_iff {w₁ w₂ : Fin (n + 1) → R} (hw₁ : ∑ j, w₁ j = 1)
+    (hw₂ : ∑ j, w₂ j = 1) {i : Fin (n + 1)} :
+    (affineSpan R (Set.range (s.faceOpposite i).points)).SSameSide
+      (Finset.univ.affineCombination R s.points w₁)
+      (Finset.univ.affineCombination R s.points w₂) ↔
+        SignType.sign (w₁ i) = SignType.sign (w₂ i) ∧ w₁ i ≠ 0 := by
+  refine ⟨fun h ↦ ?_, fun ⟨hs, h0⟩ ↦ s.sSameSide_affineSpan_faceOpposite_of_sign_eq hw₁ hw₂ hs h0⟩
+  have h0 : w₁ i ≠ 0 := by
+    intro h0
+    apply h.left_notMem
+    rw [range_faceOpposite_points]
+    exact affineCombination_mem_affineSpan_image hw₁ (by simp [h0]) s.points
+  refine ⟨?_, h0⟩
+  have h0' : w₂ i ≠ 0 := by
+    intro h0'
+    apply h.right_notMem
+    rw [range_faceOpposite_points]
+    exact affineCombination_mem_affineSpan_image hw₂ (by simp [h0']) s.points
```

Comment:
Idem dito.

Reply:
I've opened #32150 to add a lemma `affineCombination_mem_affineSpan_faceOpposite_iff`, made this PR depend on that one and refactored all the proofs here that used `range_faceOpposite_points` to use that new lemma. (The remaining case of duplication you identified will require further definitions and lemmas elsewhere to resolve.)